### ISSUE TITLE
Using Update Service Fabric App Versions Task with mutiple Applications does not work

### ISSUE THREAD

**Issue by Eneuman** [Microsoft/vsts-tasks#4389]
Hi
I have created a solution with two service fabric applications in it.
When I created a new build definition using the "Azure Service Fabric Application" (in VSTS) it sets up a build task with the name "Build solution **\*.sfproj" and a bit below this, a task with the name "Update Service Fabric App Versions".

The build runs successfully but only the last application is getting the version updated.
Looking through the build log I can see that the task "Build solution **\*.sfproj" is actually clearing the directory between each build.

> 2017-05-18T20:28:58.6704967Z Build started 5/18/2017 8:28:58 PM.
> 2017-05-18T20:28:59.4465134Z Project "d:\a\3\s\Main\src\WebApi\WebApi.sfproj" on node 1 (Package target(s)).
> 2017-05-18T20:28:59.4475134Z CleanApplicationPackage:
> 2017-05-18T20:28:59.4475134Z   rmdir /Q /S "d:\a\3\a\applicationpackage"
> 2017-05-18T20:29:00.0135241Z Project "d:\a\3\s\Main\src\WebApi\WebApi.sfproj" (1) is building "d:\a\3\s\Main\src\WebApi.Service\WebApi.Service.csproj" (2:2) on node 1 (default targets).

So once it reaches the task "Update Service Fabric App Versions", there is only one project to version.

Is this a bug or does the build definition only support building one application per solution?

As a workaround, I tried to create two seperated .sfproj build tasks with the MSBuild Argument set to

`/t:Package /p:PackageLocation=$(build.artifactstagingdirectory)\applicationpackage\<MyApplicationName>`

But then the Update versioning task failed with:

> 2017-05-22T21:01:10.6739910Z Searching for application package: d:\a\3\a\applicationpackage
> 2017-05-22T21:01:10.8669739Z Found application package: d:\a\3\a\applicationpackage
> 2017-05-22T21:01:11.1700750Z ##[error]Cannot find path 'D:\a\3\a\applicationpackage\ApplicationManifest.xml' because it does not exist.
> 2017-05-22T21:01:11.2260693Z ##[section]Finishing: Update Service Fabric App Versions - MyApplication

Is there a correct workaround for this ?

**Comment by bishal-pdMSFT** [comment]
@Eneuman the service fabric build definition template is designed to work for a sln containing just one sfproj. You need to add another "Update Service Fabric App Versions" task for second sfproj application. Additionally you should provide full path to ApplicationManifest.xml to this task. So, for two tasks, you will use paths like $(build.artifactstagingdirectory)\applicationpackage\**application1** and $(build.artifactstagingdirectory)\applicationpackage\**application2** respectively. 

Let me know if this does not solve your problem.

**Comment by Eneuman** [comment]
@bishal-pdMSFT I tried your suggestions and changed the first build task build argument to:
`/t:Package /p:PackageLocation=$(build.artifactstagingdirectory)\applicationpackage\Authentication`

and the argument on the second build task to:
`/t:Package /p:PackageLocation=$(build.artifactstagingdirectory)\applicationpackage\Ekonomi`

I also changed the versioning task, Application Package to:
`$(build.artifactstagingdirectory)\applicationpackage\Authentication`

but it still fails with the following error:
```
2017-05-24T14:24:13.3909434Z Progress update: '18464761' bytes received...
2017-05-24T14:24:13.3939454Z Previous build location: 'd:\a\tmp\drop'
2017-05-24T14:24:13.3949443Z Searching application type 'AuthenticationType' for changes...
2017-05-24T14:24:13.6099759Z ##[error]Cannot find path 'D:\a\tmp\drop\applicationpackage\Authentication\ApplicationManifest.xml' because it does not exist.
2017-05-24T14:24:13.7579540Z ##[section]Finishing: Update Service Fabric App Versions - Authentication
```

**Comment by bishal-pdMSFT** [comment]
@Eneuman I just tried this scenario locally on my box and I don't see this issue. if you are using private agent, can you please check the contents of d:\a\tmp\drop?
Also please enable debug logging by setting a build variable with name "system.debug" and value "true" and retry the release and then send me the logs

**Comment by Eneuman** [comment]
@bishal-pdMSFT I'm using Hosted Agent VS 2017 when building.
I have attached the logs below.

The only task I have added to the build definition is "dotnet restore" because I have projects using .Net Standard 1.4.
[logs_46.zip](https://github.com/Microsoft/vsts-tasks/files/1026554/logs_46.zip)

**Comment by bishal-pdMSFT** [comment]
@Eneuman I think the issue is happening when task is trying to compare new version against last successful release. For you earlier releases, you were putting manifest files directly under applicationpackage folder. And then to accommodate multiple applications, you added an extra folder under applicationpackage. And now the comparison logic is looking for that extra folder in old artifacts - which does not exist.
To make sure this is the issue, download last build using this url: 'https://caresoftab.visualstudio.com/538c4e9f-82ae-4321-a9bc-d34bd5c16cec/_apis/build/builds/36/artifacts?artifactName=drop&%24format=zip'. You should notice that there is no 'Authentication' or 'Economi' folder - which is logged as error -  "Cannot find path 'D:\a\tmp\drop\applicationpackage\Authentication\ApplicationManifest.xml' because it does not exist."

**Comment by bishal-pdMSFT** [comment]
one workaround I can think of is - create a release with "Update only if changed" flag turned off. This will not try to compare against old release and then you will have a successful release with new folder structure. Then you can enable this flag back for subsequent releases.

**Comment by Eneuman** [comment]
@bishal-pdMSFT Okey, Im gonna give it a try right now. But shouldn't the compare just log if it cannot find it in the previous release and not fail?

**Comment by bishal-pdMSFT** [comment]
yes, this can be treated as a bug.

**Comment by Eneuman** [comment]
@bishal-pdMSFT I just got the result from testing your workaround.
On the first run the build succeded and the Packages got updated to version "1.0.0.20170524.3"

On the second run (now with "only update if changed" turned on), it updated it successfuly, so the workaround works fine :)

I  did noticed a stange things tough.
The second "Update Service Fabric Version" downloaded the drop again. Shouldn't it check if it has already been downloaded and not download it again ? Maby save the downloaded file with a hash of the url it used to download it, and if the the file exists, down download it?
